src/Sparrow/Json/BlittableJsonReaderObject.cs

```diff
@@ -483,5 +485,180 @@ public void CopyTo(byte* ptr)
+
+            if (current != currentSize)
+                throw new InvalidDataException("Root metadata not valid");
+        }
+
+        private int PropertiesNamesValidation(int numberOfProps, int propsOffsetList, int propsNamesOffsetSize,
+            int currentSize)
+        {
+            var offsetCounter = 0;
+            for (var i = numberOfProps; i > 0; i--)
+            {
+                int stringLength;
+                var nameOffset = 0;
+                nameOffset = ReadNumber((_mem + propsOffsetList + i * propsNamesOffsetSize),
+                    propsNamesOffsetSize);
+                if ((nameOffset > currentSize) || (nameOffset < 0))
+                    throw new InvalidDataException("Properties names offset not valid");
+                stringLength = StringValidation(propsOffsetList - nameOffset);
+                if (offsetCounter + stringLength + 1 != nameOffset)
+                    throw new InvalidDataException("Properties names offset not valid");
+                offsetCounter = nameOffset;
+                currentSize -= (stringLength + 1);
+            }
+            return currentSize;
+        }
+
+        private int StringValidation(int stringOffset)
+        {
+            byte offset;
+            int stringLength;
+            stringLength = ReadVariableSizeInt(stringOffset, out offset);
+            if (stringLength < 0)
+                throw new InvalidDataException("String not valid");
+            var str = stringOffset + offset;
+            var stringEnd = ReadVariableSizeInt(stringOffset + offset + stringLength, out offset);
+            if (stringEnd != 0)
+            {
+                for (var i = 0; i < stringEnd; i++)
+                {
+                    var escCharOffset = ReadNumber(_mem + str + stringLength + offset + i, 1);
+                    var escChar = (char)ReadNumber(_mem + str + escCharOffset, 1);
+                    if (!(EscapeChars.Contains(escChar)))
+                        throw new InvalidDataException("String not valid");
+                }
+            }
+            return stringLength + offset + stringEnd;
+        }
+
+        private BlittableJsonToken TokenValidation(byte tokenStart, out int propOffsetSize,
+            out int propIdSize)
+        {
+            var token = (BlittableJsonToken)tokenStart;
+            var tokenType = ProcessTokenTypeFlags(token);
+            propOffsetSize = ((tokenType == BlittableJsonToken.StartObject) ||
+                               (tokenType == BlittableJsonToken.StartArray))
+                ? ProcessTokenOffsetFlags(token)
+                : 0;
+
+            propIdSize = (tokenType == BlittableJsonToken.StartObject)
+                ? ProcessTokenPropertyFlags(token)
+                : 0;
+            return tokenType;
+        }
+
+        private int PropertiesValidation(BlittableJsonToken rootTokenTypen, int mainPropOffsetSize, int mainPropIdSize,
+            int objStartOffset, int numberOfPropsNames)
```

Comment:
But you are iterating over those properties for all the objects, including internal ones that may not have the same set of properties.